src/shared/contexts/editor/index.tsx

```diff
@@ -106,74 +114,119 @@ export const EditorProvider: FC = ({children}) => {
-    (
-      rawFn: FluxToolbarFunction | FluxFunction,
-      cbParentOnUpdateText: (t: string) => void
-    ): void => {
-      const fn =
-        CLOUD && isFlagEnabled('fluxDynamicDocs')
-          ? getFluxExample(rawFn as FluxFunction)
-          : (rawFn as FluxFunction)
-
-      const text = isPipeTransformation(fn)
-        ? `  |> ${fn.example.trimRight()}`
-        : `${fn.example.trimRight()}`
-
-      const header = generateImport(fn as FluxFunction, editor.getValue())
-
-      const type =
-        isPipeTransformation(fn) || functionRequiresNewLine(fn.name)
-          ? InjectionType.OnOwnLine
-          : InjectionType.SameLine
-
-      const options = {
-        text,
-        type,
-        header,
-        triggerSuggest: true,
-        cbParentOnUpdateText,
-      }
-      inject(options)
+  const injectRaw = useCallback(
+    (injection: RawInjection) => {
+      _inject({
+        text: injection.text,
+        type: InjectionMode.SameLine,
+        triggerSuggest: false,
+      })
     },
-    [inject]
+    [_inject]
   )
 
-  const injectVariable = useCallback(
-    (variableName: string, cbToParent: (t: string) => void) => {
-      if (!editor) {
-        return null
-      }
+  const injectFunction = (injection: FunctionInjection) => {
+    if (!editor) {
+      return
+    }
+
+    const fn =
+      CLOUD && isFlagEnabled('fluxDynamicDocs')
+        ? getFluxExample(injection.function as FluxFunction)
+        : (injection.function as FluxFunction)
+
+    const text = isPipeTransformation(fn)
+      ? `  |> ${fn.example.trimRight()}`
+      : `${fn.example.trimRight()}`
+
+    const header = generateImport(fn as FluxFunction, editor.getValue())
 
-      const options = {
-        text: `v.${variableName}`,
-        type: InjectionType.SameLine,
+    const type =
+      isPipeTransformation(fn) || functionRequiresNewLine(fn.name)
+        ? InjectionMode.OnOwnLine
+        : InjectionMode.SameLine
+
+    _inject({
+      text,
+      type,
+      header,
+      triggerSuggest: true,
+    })
+  }
+
+  const injectVariable = useCallback(
+    (injection: VariableInjection) => {
+      _inject({
+        text: `v.${injection.variable}`,
+        type: InjectionMode.SameLine,
         triggerSuggest: false,
-        cbParentOnUpdateText: cbToParent,
-      }
-      inject(options)
+      })
+    },
+    [_inject]
+  )
+
+  const injectSecret = useCallback(
+    (injection: SecretInjection) => {
+      _inject({
+        text: `secrets.get(key: "${injection.secret.id}") `,
+        type: InjectionMode.SameLine,
+        header: `import "influxdata/influxdb/secrets"`,
+      })
     },
-    [editor, inject]
+    [_inject]
   )
 
+  const _setEditor = (nextEditor: EditorType) => {
+    nextEditor.getModel().onDidChangeContent(() => {
+      if (!editor) {
+        return
+      }
+      const val = editor.getValue()
+      updateCBs.current.forEach(cb => cb(val))
+    })
+
+    setEditor(nextEditor)
+  }
```

Comment:
big fan of making the editor responsible for editor-related issues

Reply:
Me too. Optional provider is way more flexible than prop drilling. Of note, the code above this isn't actually used anywhere, but denise was using a callback structure instead of basing it off the editor state updating for some reason. That stuff seemed to just be double calling the updateText functions instead of picking up on some dropped state change like i originally thought, so it melted away as i tightened up the interfaces, but I wanted to keep the option there in case I missed a special condition